doc/development/autograph.rst

```diff
@@ -0,0 +1,743 @@
+a JAX array:
+
+>>> @qml.qnode(dev)
+... def f():
+...     x = jnp.array([0.1, 0.2, 0.3])
+...     for i in range(3):
+...         qml.RX(x[i], wires=i)
+...     return qml.expval(qml.PauliZ(0))
+>>> plxpr = make_plxpr(f)()
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts)
+[Array(0.99500417, dtype=float64)]
+
+If the object you are indexing **cannot** be converted to a JAX array, it is not possible for AutoGraph to capture this for loop.
+
+If you are updating elements of the array, this must be done using the JAX ``.at`` and ``.set`` syntax.
+
+>>> def f():
+...     my_list = jnp.empty(2, dtype=int)
+...     for i in range(2):
+...         my_list = my_list.at[i].set(i)  # not my_list[i] = i
+...     return my_list
+>>> plxpr = make_plxpr(f)()
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts)
+Array([0, 1], dtype=int64)
+
+
+Dynamic indexing
+~~~~~~~~~~~~~~~~
+
+Indexing into arrays where the for loop has **dynamic bounds** (that is, where
+the size of the loop is set by a dynamic runtime variable) will also work, as long
+as the object indexed is a JAX array:
+
+>>> @qml.qnode(dev)
+... def f(n):
+...     x = jnp.array([0.0, 1 / 4 * jnp.pi, 2 / 4 * jnp.pi])
+...     for i in range(n):
+...         qml.RY(x[i], wires=0)
+...     return qml.expval(qml.PauliZ(0))
+>>> plxpr = make_plxpr(f)(0)
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, 2)
+Array(0.70710678, dtype=float64)
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, 3)
+Array(-0.70710678, dtype=float64)
+
+However AutoGraph conversion will fail if the object being indexed by the
+loop with dynamic bounds is **not** a JAX array, because you cannot index
+standard Python objects with dynamic variables. Ensure that all objects that
+are indexed within dynamic for loops are JAX arrays.
+
+Break and continue
+~~~~~~~~~~~~~~~~~~
+
+Within a for loop, control flow statements ``break`` and ``continue``
+are not currently supported.
+
+
+Updating and assigning variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For loops that update variables can also be converted with AutoGraph:
+
+>>> def f(x):
+...     for y in [0, 4, 5]:
+...         x = x + y
+...     return x
+>>> plxpr = make_plxpr(f)(0)
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, 3)
+[Array(12, dtype=int64)]
+
+However, like with conditionals, a similar restriction applies: variables
+which are updated across iterations of the loop must have a JAX compilable
+type (Booleans, Python numeric types, and JAX arrays).
+
+You can also utilize temporary variables within a for loop:
+
+>>> def f(x):
+...     for y in [0, 4, 5]:
+...         c = 2
+...         x = x + y * c
+...     return x
+>>> plxpr = make_plxpr(f)(0)
+>>> eval_jaxpr(plxpr.jaxpr, plxpr.consts, 4)
+[Array(22, dtype=int64)]
+
+Temporary variables used inside a loop --- and that are **not** passed to a
+function within the loop --- do not have any type restrictions.
+
+While loops
+-----------
+
+Most ``while`` loop constructs will be properly captured and compiled by
+AutoGraph:
```

Comment:
What about indexing within `while` loops?

Reply:
Good point! Pretty much identical to for loop, but I'll add a corresponding section here, so that the information is easy to find for each type of control flow.